File: composeApp/src/commonMain/kotlin/kr/hhp227/storygroup/ui/screens/group/GroupFeedViewModel.kt
package kr.hhp227.storygroup.ui.screens.group

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.cash.paging.PagingData
import app.cash.paging.cachedIn
import app.cash.paging.filter
import app.cash.paging.map
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kr.hhp227.storygroup.shared.domain.model.Post
import kr.hhp227.storygroup.shared.domain.usecase.GetGroupPostsPagingDataUseCase
import kr.hhp227.storygroup.shared.domain.usecase.ObservePostDeletionsUseCase
import kr.hhp227.storygroup.shared.domain.usecase.ObservePostUpdatesUseCase
import kr.hhp227.storygroup.shared.domain.usecase.ObserveUserBlocksUseCase
import kr.hhp227.storygroup.shared.domain.usecase.TogglePostLikeUseCase
import kr.hhp227.storygroup.ui.mvi.MviViewModel
import org.jetbrains.compose.resources.getString
import storygroup.composeapp.generated.resources.Res
import storygroup.composeapp.generated.resources.post_error_like

/**
 * 소식 탭 — 레거시 PostFragment의 VM 미러(탭별 VM 분리). 피드는 UiState에 담기는 최신
 * PagingData(Paging-CRUD 샘플 패턴). 갱신은 화면이 프레젠터 refresh()로 수행하므로
 * 이벤트가 없다. 수정/차단/삭제는 재조회 대신 현재 스냅샷에서 그 항목만 패치한다.
 * iosApp GroupFeedViewModel.swift와 1:1 미러
 */
class GroupFeedViewModel(
    val groupId: Long,
    getGroupPostsPagingDataUseCase: GetGroupPostsPagingDataUseCase,
    observePostUpdatesUseCase: ObservePostUpdatesUseCase,
    observeUserBlocksUseCase: ObserveUserBlocksUseCase,
    observePostDeletionsUseCase: ObservePostDeletionsUseCase,
    private val togglePostLikeUseCase: TogglePostLikeUseCase
) : ViewModel(), MviViewModel<GroupFeedViewModel.UiState, GroupFeedViewModel.Action, Nothing> {
    private val _uiState = MutableStateFlow(UiState())
    override val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    override val event: Flow<Nothing> = emptyFlow()

    private fun setPagingData(pagingData: PagingData<Post>) {
        _uiState.update { it.copy(pagingData = pagingData) }
    }

    /**
     * 수정된 게시글을 현재 스냅샷에서 그 항목만 갈아끼운다 — refresh를 태우면 첫 페이지부터
     * 전체 재조회라 이미 쌓아둔 페이지와 스크롤 위치를 잃는다(수정은 목록 구조를 바꾸지 않는다).
     * 다음 세대(새로고침·재진입)부턴 서버 값이 그대로 이긴다.
     */
    private fun applyPostUpdate(post: Post) {
        _uiState.update { state ->
            state.copy(pagingData = state.pagingData.map { if (it.id == post.id) post else it })
        }
    }

    /** 차단한 작성자의 글을 현재 스냅샷에서 걷어낸다(멤버 목록은 멤버 탭 Refresh가 걸러낸다) */
    private fun removeBlockedAuthorPosts(userId: Long) {
        _uiState.update { state ->
            state.copy(pagingData = state.pagingData.filter { it.userId != userId })
        }
    }

    /** 삭제된 글을 현재 스냅샷에서 걷어낸다 — 다음 세대부턴 서버 응답에 애초에 없다 */
    private fun removeDeletedPost(postId: Long) {
        _uiState.update { state ->
            state.copy(pagingData = state.pagingData.filter { it.id != postId })
        }
    }

    override fun onAction(action: Action) {
        when (action) {
            is Action.ToggleLike -> toggleLike(action.post)
            Action.DismissLikeError -> _uiState.update { it.copy(likeError = null) }
        }
    }

    /** 성공 반영은 리포지토리의 postUpdates 알림(applyPostUpdate)이 담당 — 여기선 실패만 다룬다 */
    private fun toggleLike(post: Post) {
        viewModelScope.launch {
            try {
                togglePostLikeUseCase(post.groupId, post.id, !post.likedByMe)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                _uiState.update { it.copy(likeError = e.message ?: getString(Res.string.post_error_like)) }
            }
        }
    }

    init {
        // UseCase는 cachedIn 없는 Flow를 반환하므로 프레젠테이션 경계인 여기서 캐시를 적용한다
        getGroupPostsPagingDataUseCase(groupId)
            .cachedIn(viewModelScope)
            .onEach(::setPagingData)
            .launchIn(viewModelScope)
        // 상세 화면에서 수정하면 목록도 바뀐 본문을 보여야 한다 — 재조회 대신 그 항목만 교체
        observePostUpdatesUseCase()
            .onEach(::applyPostUpdate)
            .launchIn(viewModelScope)
        // 차단하면 그 사람의 글이 목록에서 사라져야 한다 — 재조회 대신 그 항목들만 제거
        observeUserBlocksUseCase()
            .onEach(::removeBlockedAuthorPosts)
            .launchIn(viewModelScope)
        // 상세에서 삭제하면 목록에서도 사라져야 한다 — 재조회 대신 그 항목만 제거
        observePostDeletionsUseCase()
            .onEach(::removeDeletedPost)
            .launchIn(viewModelScope)
    }

    data class UiState(
        val pagingData: PagingData<Post> = PagingData.empty(),
        // 카드 좋아요 실패 안내 — 서버 확정 방식이라 실패해도 되돌릴 UI 상태가 없다
        val likeError: String? = null
    )

    sealed interface Action {
        data class ToggleLike(val post: Post) : Action
        data object DismissLikeError : Action
    }
}

File: iosApp/iosApp/UI/Screens/Group/GroupFeedViewModel.swift
import Combine
import Foundation
import Shared

/// 소식 탭 — composeApp GroupFeedViewModel.kt와 1:1 미러(탭별 VM 분리, 레거시 탭 Fragment VM 구조).
/// 피드는 UiState에 담기는 최신 PagingData. 갱신은 VM이 스트림을 통째로 갈아끼우는 방식이다 —
/// Compose 쪽 상세는 NavHost 목적지라 글쓰기 복귀 때 화면이 새로 만들어지는데, 그 첫 프레임엔
/// 일회성 Event도 프레젠터 refresh()도 닿지 못한다. 트리거는 상태라서 타이밍과 무관하다.
final class GroupFeedViewModel: MviViewModel {
    typealias Event = Never

    @Published private(set) var uiState = UiState()

    // 스트림을 통째로 갈아끼우는 트리거 — 복귀 첫 프레임에 유실되지 않도록 이벤트가 아닌 상태다
    private let refreshTrigger = CurrentValueSubject<Int, Never>(0)

    let groupId: Int64

    // 목록 카드용 좋아요 토글 — 상세용 setPostLikedUseCase와 달리 좋아요 목록을 다시 읽지 않는다
    private let togglePostLikeUseCase: TogglePostLikeUseCase

    private var cancellables = Set<AnyCancellable>()

    private func setPagingData(_ pagingData: PagingData<Post>) {
        uiState.pagingData = pagingData
    }

    /// 수정/차단/삭제는 재조회 대신 현재 스냅샷에서 그 항목만 패치한다(Kotlin과 동일 —
    /// suspend 변환이라 Swift 클로저를 못 넘겨 PostBridges 브리지 함수 사용)
    private func applyPostUpdate(_ post: Post) {
        uiState.pagingData = PostBridgesKt.postPagingDataWithUpdate(pagingData: uiState.pagingData, post: post)
    }

    private func removeBlockedAuthorPosts(_ userId: Int64) {
        uiState.pagingData = PostBridgesKt.postPagingDataWithoutAuthor(pagingData: uiState.pagingData, userId: userId)
    }

    private func removeDeletedPost(_ postId: Int64) {
        uiState.pagingData = PostBridgesKt.postPagingDataWithoutPost(pagingData: uiState.pagingData, postId: postId)
    }

    func onAction(_ action: Action) {
        switch action {
        // 글쓰기 성공·당겨서 새로고침 — 새 PagingSource가 첫 페이지부터 다시 읽는다
        case .refresh: refreshTrigger.send(refreshTrigger.value + 1)
        case .toggleLike(let post): toggleLike(post)
        case .dismissLikeError: uiState.likeError = nil
        }
    }

    /// 성공 반영은 리포지토리의 postUpdates 알림(applyPostUpdate)이 담당 — 여기선 실패만 다룬다
    private func toggleLike(_ post: Post) {
        Task { @MainActor in
            do {
                try await togglePostLikeUseCase.invoke(groupId: post.groupId, postId: post.id, liked: !post.likedByMe)
            } catch {
                uiState.likeError = error.kotlinMessage(fallback: "좋아요 처리에 실패했습니다.")
            }
        }
    }

    init(
        groupId: Int64,
        getGroupPostsPagingDataUseCase: GetGroupPostsPagingDataUseCase = AppContainer.shared.getGroupPostsPagingDataUseCase,
        observePostUpdatesUseCase: ObservePostUpdatesUseCase = AppContainer.shared.observePostUpdatesUseCase,
        observeUserBlocksUseCase: ObserveUserBlocksUseCase = AppContainer.shared.observeUserBlocksUseCase,
        observePostDeletionsUseCase: ObservePostDeletionsUseCase = AppContainer.shared.observePostDeletionsUseCase,
        togglePostLikeUseCase: TogglePostLikeUseCase = AppContainer.shared.togglePostLikeUseCase
    ) {
        self.groupId = groupId
        self.togglePostLikeUseCase = togglePostLikeUseCase

        // UseCase는 cachedIn 없는 Flow를 반환하므로 프레젠테이션 경계인 여기서 캐시를 적용한다
        // (Kotlin: refreshTrigger.flatMapLatest { useCase(groupId) }.cachedIn(viewModelScope))
        refreshTrigger
            .map { _ in getGroupPostsPagingDataUseCase(groupId: groupId).cachedIn() }
            .switchToLatest()
            .sink { [weak self] in self?.setPagingData($0) }
            .store(in: &cancellables)
        KotlinFlowPublisher<Post> { onEach in
            observePostUpdatesUseCase.updatesFlow().subscribe(onEach: onEach)
        }
        .sink { [weak self] post in self?.applyPostUpdate(post) }
        .store(in: &cancellables)
        KotlinFlowPublisher<KotlinLong> { onEach in
            observeUserBlocksUseCase.blocksFlow().subscribe(onEach: onEach)
        }
        .sink { [weak self] userId in self?.removeBlockedAuthorPosts(userId.int64Value) }
        .store(in: &cancellables)
        KotlinFlowPublisher<KotlinLong> { onEach in
            observePostDeletionsUseCase.deletionsFlow().subscribe(onEach: onEach)
        }
        .sink { [weak self] postId in self?.removeDeletedPost(postId.int64Value) }
        .store(in: &cancellables)
    }

    struct UiState {
        var pagingData: PagingData<Post> = PostBridgesKt.emptyPostPagingData()
        // 카드 좋아요 실패 안내 — 서버 확정 방식이라 실패해도 되돌릴 UI 상태가 없다
        var likeError: String? = nil
    }

    enum Action {
        case refresh
        case toggleLike(Post)
        case dismissLikeError
    }
}
